core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -572,20 +572,35 @@ private[spark] class ExecutorAllocationManager(
   }
 
   /**
+   * Callback invoked when an extra executor is needed (See SPARK-19326)
+   */
+  private def onExtraExecutorNeeded(): Unit = synchronized {
+    val maxNeeded = math.max(math.min(maxNumExecutorsNeeded + 1, maxNumExecutors), minNumExecutors)
```

Comment:
Perhaps what we need is just to ensure there are more than two active executors left, so we may meet the locality requirement and launch the speculative jobs?